orchestrator/graphql/add_graphql.py

```diff
@@ -0,0 +1,105 @@
+
+
+def create_strawberry_enum(enum: Any) -> EnumMeta:
+    return strawberry.enum(enum)
+
+
+def is_not_strawberry_enum(key: str, strawberry_enums: EnumList) -> bool:
+    return key not in strawberry_enums
+
+
+def is_enum(field: Any) -> bool:
+    return inspect.isclass(field) and issubclass(field, Enum)
+
+
+def graphql_name(name: str) -> str:
+    return to_camel(name.replace(" ", "_").replace("Initial", "").replace("Inactive", ""))
+
+
+def graphql_subscription_name(name: str) -> str:
+    return f"{graphql_name(name)}Subscription"
+
+
+def is_not_strawberry_type(key: str, strawberry_models: StrawberryModelType) -> bool:
+    return graphql_name(key) not in strawberry_models
+
+
+def create_subscription_strawberry_type(strawberry_name: str, model: Type[DomainModel]) -> Type[SubscriptionInterface]:
+    base_type = type(strawberry_name, (SubscriptionInterface,), {})
+    pydantic_wrapper = strawberry.experimental.pydantic.type(model, all_fields=True)
+    federation_wrapper = strawberry.federation.type(description=f"{strawberry_name} Type", keys=["subscriptionId"])
+    pydantic_type = pydantic_wrapper(base_type)
+    federation_type = type(strawberry_name, (pydantic_type,), {})
+    strawberry_type = federation_wrapper(federation_type)
+
+    def from_pydantic(model: pydantic_type) -> strawberry_type:  # type: ignore
+        graphql_model = pydantic_type.from_pydantic(model)
+        graphql_model.__class__ = strawberry_type
+        return graphql_model
+
+    strawberry_type.from_pydantic = from_pydantic  # type: ignore
+    return strawberry_type
+
+
+def create_block_strawberry_type(
+    strawberry_name: str, model: Type[DomainModel]
+) -> Type[StrawberryTypeFromPydantic[DomainModel]]:
+    new_type = type(strawberry_name, (), {})
+    strawberry_wrapper = strawberry.experimental.pydantic.type(model, all_fields=True)
+    return strawberry_wrapper(new_type)
+
+
+def create_strawberry_enums(model: Type[DomainModel], strawberry_enums: EnumList) -> EnumList:
+    enums = {
+        key: field
+        for key, field in model._non_product_block_fields_.items()
+        if is_enum(field) and is_not_strawberry_enum(key, strawberry_enums)
+    }
+    return strawberry_enums | {key: create_strawberry_enum(field) for key, field in enums.items()}
+
+
+def add_class_to_strawberry(
+    model_name: str,
+    model: Type[DomainModel],
+    strawberry_models: StrawberryModelType,
+    strawberry_enums: EnumList,
+    with_interface: bool = False,
+) -> None:
+    strawberry_enums = create_strawberry_enums(model, strawberry_enums)
+
+    product_blocks_types_in_model = get_depends_on_product_block_type_list(model._get_depends_on_product_block_types())
+    for field in product_blocks_types_in_model:
+        if is_not_strawberry_type(field.__name__, strawberry_models) and field.__name__ != model_name:
+            add_class_to_strawberry(field.__name__, field, strawberry_models, strawberry_enums)
+
+    strawberry_name = graphql_name(model_name)
```

Comment:
The way that the name of a model is determined is now a bit scattered among the `register_graphql()` function (for SubscriptionModel) and here in this function (for ProductBlockModel). 

I think it would be better to only pass `model` to this function, and then let this function (or another helper function) decide the `strawberry_name` with something like a 

```py
if isinstance(model, SubscriptionModel):
    strawberry_name = graphql_subscription_name(model.__base_type__.__name__)
elif isinstance(model, ProductBlockModel):
    strawberry_name = graphql_name(model.__name__)
else:
    raise TypeError("some helpful error message")
```

What do you think?

Reply:
Also it might be helpful to add an extra check and debug logging (I'll admit I'm a bit of a logging addict..)

```py
if strawberry_name in strawberry_models:
    logger.debug("Skip already registered strawberry model", model=repr(model), strawberry_name=strawberry_name)
    return None

logger.debug("Registering strawberry model", model=repr(model), strawberry_name=strawberry_name)
```

Reply:
Alright! And what do you think about my first suggestion of only passing the DomainModel derivative into this function, and letting the name generation take place here?